### PR TITLE
Make Input threadsafe (optionally)

### PR DESCRIPTION
**Pull request by justinj**

### CODE REVIEW

File: hydroflow/src/scheduled/input.rs [hunk 1]
```diff
@@ -1,31 +1,54 @@
...
+use std::{cell::RefCell, marker::PhantomData, rc::Rc, sync::mpsc::SyncSender};
 
 use super::{OpId, Reactor};
 
-pub struct Input<T> {
+pub trait Givable<T> {
+    fn give(&self, t: T) -> bool;
+}
+
+impl<T> Givable<T> for Rc<RefCell<Vec<T>>> {
```
Comment (MingweiSamuel) [inline hydroflow/src/scheduled/input.rs:9]: Don't love the `Input<T, Rc<RefCell<Vec<T>>>>`  compared to `Input<T, SyncSender<T>>` but is ok for now. Maybe mark TODO(mignwei) if you agree?

File: hydroflow/src/scheduled/input.rs [hunk 2]
```diff
@@ -1,31 +1,54 @@
-use std::{cell::RefCell, rc::Rc};
+use std::{cell::RefCell, marker::PhantomData, rc::Rc, sync::mpsc::SyncSender};
 
 use super::{OpId, Reactor};
 
-pub struct Input<T> {
+pub trait Givable<T> {
```
Comment (MingweiSamuel) [inline hydroflow/src/scheduled/input.rs:5]: I think `Give` is the more idiomatic name btw

**Review by MingweiSamuel — APPROVED** [review]

File: hydroflow/src/scheduled/input.rs (hunk 1 above)
Comment (justinj) [reply hydroflow/src/scheduled/input.rs]: Gave it a slightly friendlier name.

File: hydroflow/src/scheduled/input.rs (hunk 2 above)
Comment (justinj) [reply hydroflow/src/scheduled/input.rs]: Done.

**Comment by justinj** [comment]
TFTR!

**Pull request merged**